Add `CONTRIBUTING.md`

liammulh opened the pull request:
- this is a draft; I need some feedback and help making this solid
- this PR adds a contributing document
- I tried to make the contributing document fairly comprehensive
- the doc has a link to a series of videos I made for people who are new to software development who want to contribute to Numberscope (I thought video would be easier to follow than text for people who are new -- that's how it was for me when I was new)
- I still need to add a "Open a pull request" section and add a link to it in the list of scenarios where it makes sense to do so

When I started work on this, I was just going to add a doc for opening pull requests, but it sort of turned into "how to contribute generally". Any/all feedback is welcome and appreciated. :)

liammulh commented:
Also, I see that I accidentally deleted the `doc/contributing.md` file. (I thought I was getting rid of a file I renamed.) Anyway, we should incorporate the content of that document into this one.

Also, I neglected to think about the docsite, so if the docsite isn't picking up markdown documents in the root of the repo, we'll need to tell it to pick this one up. (It's in the root of the repo because that's where GitHub expects the CONTRIBUTING doc to be, I think.)

katestange commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,193 @@
+
+```sh
+git add /path/to/file/here
+```
````
mention that you have to stage each file you are interested in (i.e. there could be more than one)

katestange commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,193 @@
+```
+
+## Do your work on a branch in your fork
+
````
This title sounds like a command rather than a section title.  Maybe "How to contribute code:  add, commit and push"  Then you could add a sentence that you are working on your branch in your fork.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+4. When you think your code is in a good state, submit a pull request.
+
+## Scenarios and what to do about them
+
```
I think this big flowchart could be useful for our students who got started in some particular strange path.  But it's rather daunting because it is so big.  Maybe we should have first a section that is for the "new student" who is going to do everything correctly in order the first time according to our plan.  Then, we can have the sections that explain the various different types of things to do (stash, commit, clone, whatever).  Then, you can have a flowchart at the end for all the "other" scenarios that might arise.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+2. Create a branch within that repository.
+3. Do your work on the branch you created. (Make sure your pre-commit hooks
+   are working.)
+4. When you think your code is in a good state, submit a pull request.
```
what are pre-commit hooks?  is that mentioned anywhere else?  are there important things to say?

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+Numberscope.
+
+## If you have software dev experience, here's the TL;DR
+
```
Maybe a title like "Here's the TL;DR / overview" because this is useful even if you don't have dev experience.  It will give the big picture.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+   are working.)
+4. When you think your code is in a good state, submit a pull request.
+
+## Scenarios and what to do about them
```
Maybe after the TL;DR but before the scenarios, you should put a link to a really good git overview on the right level of detail for us, so that people who want the "big git picture" can look at that.  It's something that I'm assuming is done well in various places, so we probably don't need to repeat that general rundown ourselves.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+GitHub allows you to clone the repo a few different ways:
+
+1. via HTTPS (easiest method, doesn't require setup, has limited
+   functionality)
```
does the limited functionality preclude using it for development?  If so, maybe point out we aren't using this one because of that.  Maybe "but doesn't have the functionality we need" or something along those lines.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+## Add a remote
+
+To add a remote version of your repository, issue the following command:
+
```
I'm not sure I really understand what this does.  Can you add a sentence?

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+## Create a fork
+
+Note: This is a GitHub operation, not a Git operation.
+
```
Maybe add a sentence like "it creates your own private copy of the numberscope repository that is linked to the main repository so that changes can be pushed" or something more accurate than my crude understanding I just demonstrated...

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+Note: This is a GitHub operation, not a Git operation.
+
+1. Go to the page of the repository you want to clone. For instance, if you
+   want to clone the numberscope/frontscope repository, go to
```
"you want to fork"? (the word clone appears twice)

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+## Clone a repo
+
+To clone a Git repository to your computer, issue the following command:
+
```
Again, maybe one sentence as to what this is.  Like, "cloning a repository creates a copy of it on your computer that you can modify and, ultimately, send those changes back to the original on github"

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+When using Git, it can be helpful to think of it as photographer who stages
+your work, takes a snapshot of it, and stores the photo somewhere safe (i.e.
+GitHub).
+
```
I like this sentence, and the way you echoed each part in the sections below.  That's very nice.  Maybe we can also echo forward: "... who stages your work ("adding" changes), takes a snapshot of it ("committing" changes), and store the photo somewhere safe, i.e. GitHub ("pushing" changes)."  That way, it connects slightly more to the next sections.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+                -   [Do your work on a branch in your fork](#do-your-work-on-a-branch-in-your-fork).
+
+## Stash your changes
+
```
All these sections here (one for each command) are so clean and clear and crisp, it's very easy to navigate.  I like this style.

katestange commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,193 @@
+This video series will introduce you to Numberscope, GitHub, the software we
+use in the Numberscope project, and the workflow for contributing to
+Numberscope.
+
```
I haven't watched these as part of the review, I figured I'd start with the doc for now.  But it's great that you made them!  What an excellent resource to have!

katestange reviewed (commented):
I just gave it a quick go over since it sounded like you were thinking of this as a first draft.  Looking good!

liammulh commented:
Thanks! I will read and incorporate the feedback soon.

gwhitney commented:
Absolutely mkdocs should be picking up .md files in the top level directory. If you have any problem with it picking this up let me know and I will grab a copy of the branch and take a look.

liammulh commented:
I've added a lot of changes in a new branch that also adds:

- new docs in the doc directory
- some images

I also changed the contributing doc to address @katestange's suggestions/feedback above.

I will submit a new PR for the new branch and close this PR. Thanks for the feedback.
